### PR TITLE
Implement deterministic e-cash notes

### PR DESCRIPTION
**Pull request by elsirion**
Working towards #502.

TODO:
- [x] spec out key derivation: [rendered document](https://github.com/elsirion/fedimint/blob/2022-10-deterministic-ecash/docs/recoverable_e-cash.md) *Less fancy now, without graphics, but actually implementing it showed me some ergonomic weaknesses with the old scheme and the updated one was too complex for one big graphic. If anyone wants to give it a shot feel free 😅*
- [x] implement key derivation
- [ ] ~investigate private information retrieval scheme to retrieve signatures~ (possibly separate PR)

I think the actual recovery should become a separate PR. Some ideas around it:

* Go with a simple "fetch all blind sigs+spends since last daily backup" approach first. This should be plenty efficient for low volume federations and doesn't allow the federation to correlate transactions.
* Later on a private information retrieval scheme should be used, but more research is needed for that.

### CODE REVIEW

**Comment by elsirion** [comment]
I'm getting some advice not to use BIP39, I don't have a strong opinion on the encoding of our input key material, so I'm open to other ideas. Current contenders are [SLIP39](https://github.com/satoshilabs/slips/blob/master/slip-0039.md) and [Codex32](https://github.com/roconnor-blockstream/SSS32/blob/ms32/MasterSeed32.md).

For now I'll not worry about it too much and assume a suitable pseudo random key to be supplied no matter the encoding, if anyone feels strongly about the encoding please chime in.

**Review by dpc — APPROVED** [review]
Read it, LGTM.

**Comment by dpc** [comment]
> I'm getting some advice not to use BIP39

[BIP39 is kind of meh](https://github.com/rust-bitcoin/rust-bitcoin/discussions/814#discussion-3838325), but since in Fedimint the nonces are one time use only, it should be very practical to phase out usage of any previous format and migrate to a new key material using new key format. Unlike in Bitcoin where the addresses are kind of forever, in Fedimint we can always just remint all unspent notes to nonces newly derived from a new key, let the user back it up again and throw away the old one.

**Comment by dpc** [comment]
@elsirion BTW. My previous comment made me think. Maybe we should introduce an additional derivation element that would mean "generation". This could let us avoiding deriving long sequences of nonces, where 99.9% of them would be already spent, by periodically switching to a new "generation".

Checking which generation to use is simple and efficient: initially just keep trying key 0 in every generation, until you hit unused one. The last used generation is the current one, and we are guaranteed not to have any unspent notes on any generation other than last and one previous one (in case something failed in between moving things to a new generation).

File: crypto/hkdf/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+/// Implements the [RFC5869] hash based key derivation function using the hash function `H`.
+///
+/// [RFC5869]: https://www.rfc-editor.org/rfc/rfc5869
+struct HKDF<H: BitcoinHash> {
+    prk: Hmac<H>,
+}
+
+impl<H: BitcoinHash> HKDF<H> {
+    /// Runs HKDF-extract and keeps the resulting pseudo random key as internal state
```
Comment (dpc) [inline crypto/hkdf/src/lib.rs:30]: style: Imperative mood is more idiomatic.

**Comment by elsirion** [comment]
> Maybe we should introduce an additional derivation element that would mean "generation".

I think daily backups are an easier solution, then you only have to search for notes since that checkpoint. Generations would imply re-issuing all the notes regularly, which might be a privacy risk. If you can trigger some networked code already every now and then, then you can as well save a backup at a predetermined location (federation or OS provided backup location like google cloud, if it's encrypted it doesn't matter all that much).

File: client/client-lib/src/mint/mod.rs [hunk 2]
```diff
@@ -101,6 +105,28 @@ impl<'c> MintClient<'c> {
             .collect()
     }
 
+    fn new_note_secret(&self, dbtx: &mut DatabaseTransaction<'_>) -> DerivableSecret {
```
Comment (dpc) [inline client/client-lib/src/mint/mod.rs:108]: We should have a test that runs this in parallel for many threads in loop and verify that it's atomic.

What's consequence of re-using same nonce twice? Any bad consequences of skipping a nonce?

File: client/client-lib/src/mint/mod.rs [hunk 3]
```diff
@@ -241,39 +268,13 @@ impl<'c> MintClient<'c> {
...
-            coins = %requests.item_count(),
-            tiers = ?requests.tiers().collect::<Vec<_>>(),
-            "Generated issuance request"
-        );
-
-        let sig_req = SignRequest(blinded_nonces);
-        let issuance_req = NoteIssuanceRequests { coins: requests };
-
-        (issuance_req, sig_req)
+impl Extend<(Amount, NoteIssuanceRequest)> for NoteIssuanceRequests {
```
Comment (dpc) [inline client/client-lib/src/mint/mod.rs:271]: TIL: `Extend`

**Review by dpc — APPROVED** [review]
One thought/question about atomicity, but generally LGTM.

File: client/client-lib/src/mint/mod.rs (hunk 2 above)
Comment (elsirion) [reply client/client-lib/src/mint/mod.rs:108]: Good idea!

Consequences:
* Reuse: burning money
* Skipping: need to account for that when recovering (gap limit)

File: client/client-lib/src/mint/mod.rs (hunk 3 above)
Comment (elsirion) [reply client/client-lib/src/mint/mod.rs:271]: It's needed for `unzip`, confuses me every time that `FromIterator` doesn't work for that.

File: client/client-lib/src/mint/mod.rs (hunk 2 above)
Comment (dpc) [reply client/client-lib/src/mint/mod.rs:108]: #861 created, as I just landed this PR

Comment (elsirion) [reply client/client-lib/src/mint/mod.rs:108]: Really good instinct! Had to fix a bug with rocksdb isolation c8597b644451cfc1c1955c7b0455784c6847ab3a. @m1sterc001guy did we just miss this or should we not need snapshot isolation and is that even the right term for it?

**Comment by elsirion** [comment]
Oh lol, you already merged it

**Pull request merged**